Disable changing Custom RTC when game is running

RisingFog opened the pull request:
Because I merged #4007 too early, I missed out on a glaringly obvious issue.

This should make it so you can't actually edit custom RTC via the GUI while a game is running.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4033)
<!-- Reviewable:end -->

JosJuice commented:
I would prefer if you used the approach that someone suggested on IRC. (I forgot who it was, sorry!) Instead of storing the local time at boot and using that in `GetGCTime()`, you would store `time_offset = localtime - m_customRTCValue` at boot, and then return `localtime - time_offset` from `GetGCTime()`. This means that you don't need a special case for `bEnableCustomRTC` in `GetGCTime()` – you'll just need to set time_offset to 0 at boot if `!bEnableCustomRTC`. It also means that modifications to `m_customRTCValue` impossibly can have any effect while emulation is running.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 174 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3PYJeAFaEoaxOppWZ:-KN3PYJeAFaEoaxOppW_:1863100413) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L174)):*
> ```cpp
>   wxDateTime custom_rtc(static_cast<time_t>(SConfig::GetInstance().m_customRTCValue));
>   custom_rtc = custom_rtc.ToUTC();
>   static bool custom_rtc_enabled = SConfig::GetInstance().bEnableCustomRTC;
> ```

Why? This doesn't make much sense.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 200 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3PeoIv7kDdvhJREgd:-KN3PeoJl1w1Tq7eblcr:1742353533) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L200)):*
> ```cpp
>     m_custom_rtc_date_picker->Enable(false);
>     m_custom_rtc_time_picker->Enable(false);
>   }
> ```

What if the result of `IsRunning()` changes while the window is open?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-:-KN3PrwzkCWiEksC3TAj:-214397884)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 174 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3QqFl7t-KdBwYhAzo:-KN3QqFmrYkWamQyiYxb:1055630371) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L174)):*
> ```cpp
>   wxDateTime custom_rtc(static_cast<time_t>(SConfig::GetInstance().m_customRTCValue));
>   custom_rtc = custom_rtc.ToUTC();
>   static bool custom_rtc_enabled = SConfig::GetInstance().bEnableCustomRTC;
> ```

Why is static necessary here as opposed to just letting it load that value after the game being played has ended?

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 190 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3QheeLXtM7DId6BuC:-KN3Qhefs1GZbejL7NCB:1289634194) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L190)):*
> ```cpp
>     m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Jan, 2000),
>                                        wxDateTime(31, wxDateTime::Dec, 2099));
>   if (!Core::IsRunning())
> ```

You can simplify this by handling the case where the core is running first, turning it into

```cpp
if (Core::IsRunning())
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

lioncash commented:
*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 200 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3PeoIv7kDdvhJREgd:-KN3Rxc-ueqei0fssaPN:1063341793) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L200)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> What if the result of `IsRunning()` changes while the window is open?
</details>
That's not handled by any other config settings, as far as I know. Ideally, you'd fire off a custom wx event that signifies that the core has stopped/started and handle it in appropriate event handler functions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 200 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3PeoIv7kDdvhJREgd:-KN3T7QrcWJYQcgh7c8X:-671351936) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L200)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> That's not handled by any other config settings, as far as I know. Ideally, you'd fire off a custom wx event that signifies that the core has stopped/started and handle it in appropriate event handler functions.
</details>
I'm fine with not handling it as long as the RTC is prevented from being changed in some other way, for instance the one I described.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 174 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3QqFl7t-KdBwYhAzo:-KN3VFswCNji7eU3JDm8:1740281843) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L174)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Why is static necessary here as opposed to just letting it load that value after the game being played has ended?
</details>
Removed.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 174 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3PYJeAFaEoaxOppWZ:-KN3VGjM0ZF3O9CPgqYQ:1740281843) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L174)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Why? This doesn't make much sense.
</details>
Removed.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 190 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3QheeLXtM7DId6BuC:-KN3VIVbNA1m99-RKGeH:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d57ccabf5f18cddd6bc7fe35344a725abf446755/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L190)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can simplify this by handling the case where the core is running first, turning it into
> 
> ```cpp
> if (Core::IsRunning())
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

lioncash commented:
*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 152 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3UfvvYEMbPCOgqcxV:-KN3UfvweXjk2UjZuhAR:-1966309738) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e5fef7cdf7b43ac023ce1f4145aea0d394a39912/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L152)):*
> ```cpp
>   if (Core::IsRunning())
>   {
>     m_custom_rtc_date_picker->SetValue(static_cast<time_t>(m_temp_date));
> ```

You chould probably extract the control disabling code into its own function, since it's identical across more than one place.

Perhaps something like:

```cpp

void AdvancedConfigPane::DisableCustomRTCControls()
{
    m_custom_rtc_checkbox->Enable(false);
    m_custom_rtc_date_picker->Enable(false);
    m_custom_rtc_time_picker->Enable(false);
}
```

or

```cpp

void AdvancedConfigPane::EnableCustomRTCControls(bool enable)
{
    m_custom_rtc_checkbox->Enable(enable);
    m_custom_rtc_date_picker->Enable(enable);
    m_custom_rtc_time_picker->Enable(enable);
}
```

If you want to make it applicable to both scenarios.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 87 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3aSIlDxPMvnK9TCaf:-KN3aSIlDxPMvnK9TCag:-172694971) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L87)):*
> ```cpp
> 
> // Custom RTC
> static u64 s_localtime_rtc_offset = 0;
> ```

Don't you want to use a signed type for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 233 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3ceB1sAnbX69R_sI4:-KN3ceB2sJ_9iTGRccD3:542911471) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6b9ee94555ff57c29e80950981352b9aea552f43/Source/Core/Core/HW/SystemTimers.cpp#L233)):*
> ```cpp
>   if (SConfig::GetInstance().bEnableCustomRTC)
>     s_localtime_rtc_offset =
>         Common::Timer::GetLocalTimeSinceJan1970() - SConfig::GetInstance().m_customRTCValue;
> ```

You need to add an else condition that sets it to 0. Otherwise, running one session with custom RTC followed by one without will give the user the wrong time in the second session.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3bEcHequZbSLCE-Un:-KN3bEcI3qpcHGst0J-n:-2022305188) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L232)):*
> ```cpp
>   // store and convert localtime at boot to timebase ticks
>   if (SConfig::GetInstance().bEnableCustomRTC)
>     s_localtime_rtc_offset =
> ```

This body should probably be braced, since it goes onto two lines now (still valid of course, though generally  it flows better to brace if statements that run onto two lines or more).

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 24 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3d4GLiLaQpgpqmkTL:-KN3d4GMliTqwqzPBRbc:-505363986) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6b9ee94555ff57c29e80950981352b9aea552f43/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L24)):*
> ```cpp
> {
>   InitializeGUI();
>   LoadGUIValues();
> ```

You can probably just add

```cpp
if (Core::IsRunning())
  DisableCustomRTCControls();
```

after `LoadGUIValues()` instead of checking and disabling in all the events, by the way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 233 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3ceB1sAnbX69R_sI4:-KN3dSoWJH5Be2hRfFwD:-1120268600) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6b9ee94555ff57c29e80950981352b9aea552f43/Source/Core/Core/HW/SystemTimers.cpp#L233)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> You need to add an else condition that sets it to 0. Otherwise, running one session with custom RTC followed by one without will give the user the wrong time in the second session.
</details>
Nevermind, I missed that it gets set to 0 on shutdown.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 87 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3aSIlDxPMvnK9TCaf:-KN3fUXAffHI2KhIXPR_:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L87)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Don't you want to use a signed type for this?
</details>
Done.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3bEcHequZbSLCE-Un:-KN3fVVoYhS-kfknqVSz:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L232)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This body should probably be braced, since it goes onto two lines now (still valid of course, though generally  it flows better to brace if statements that run onto two lines or more).
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 152 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3UfvvYEMbPCOgqcxV:-KN3fYXBpKRMyliHuXt8:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e5fef7cdf7b43ac023ce1f4145aea0d394a39912/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L152)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You chould probably extract the control disabling code into its own function, since it's identical across more than one place.
> 
> Perhaps something like:
> 
> ```cpp
> 
> void AdvancedConfigPane::DisableCustomRTCControls()
> {
>     m_custom_rtc_checkbox->Enable(false);
>     m_custom_rtc_date_picker->Enable(false);
>     m_custom_rtc_time_picker->Enable(false);
> }
> ```
> 
> or
> 
> ```cpp
> 
> void AdvancedConfigPane::EnableCustomRTCControls(bool enable)
> {
>     m_custom_rtc_checkbox->Enable(enable);
>     m_custom_rtc_date_picker->Enable(enable);
>     m_custom_rtc_time_picker->Enable(enable);
> }
> ```
> 
> If you want to make it applicable to both scenarios.
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 24 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3d4GLiLaQpgpqmkTL:-KN3fdfuX20EqOBcuxjh:-1233246951) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6b9ee94555ff57c29e80950981352b9aea552f43/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L24)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can probably just add
> 
> ```cpp
> if (Core::IsRunning())
>   DisableCustomRTCControls();
> ```
> 
> after `LoadGUIValues()` instead of checking and disabling in all the events, by the way.
</details>
The checks are in place in case someone starts a game while the config menu is open.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 24 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3d4GLiLaQpgpqmkTL:-KN3gKS926ogSALcQIpr:-1501980796) ([raw file](https://github.com/dolphin-emu/dolphin/blob/6b9ee94555ff57c29e80950981352b9aea552f43/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L24)):*
<details><summary><i>Previously, RisingFog (Chris Burgener) wrote…</i></summary>
> The checks are in place in case someone starts a game while the config menu is open.
</details>
Starting a game while the config window is open is very hard (or impossible?) since the config window blocks access to the main window. Checking if we need to disable GUI elements all the time isn't really worth it, especially since it's only a half-working solution (nothing gets disabled until the user tries to interact with it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Reviewed 1 of 2 files at r4, 2 of 2 files at r6, 1 of 2 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3bEcHequZbSLCE-Un:-KN3pU0MQqz8hwwXOwze:-824305348) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L232)):*
<details><summary><i>Previously, RisingFog (Chris Burgener) wrote…</i></summary>
> Done.
</details>
It doesn't look done to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 232 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033#-KN3bEcHequZbSLCE-Un:-KN3qVIEZl7ZJ0LrP7AW:-2137224273) ([raw file](https://github.com/dolphin-emu/dolphin/blob/da7daba398182a28e0c11c96f568889bc22469d5/Source/Core/Core/HW/SystemTimers.cpp#L232)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> It doesn't look done to me.
</details>
Done, for real this time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4033)*
<!-- Sent from Reviewable.io -->

Fallcrest commented:
@RisingFog Are there any instances where changing the internal clock value seriously messes up the system?

RisingFog commented:
@Fallcrest not that I'm aware of, but you're not usually able to change the system clock while running the game.
